File: csharp/MigrateSetup.cs
// Idempotent CREATE TABLE IF NOT EXISTS migrates+migrate_logs; DDL constants mirror scripts/lib/datasource-migrate.mjs:setupSql so the csharp + rust + ts runners share one migrates contract.

namespace Deterministic.MigrateRunner;

internal sealed class MigrateSetup(ISqlDialectFactory dialects, IConnectionResolver connections)
    : IMigrateCommand
{
    private static readonly string HelpText = HelpTemplates.Read("setup");

    public string Name => "setup";

    public async Task<int> RunAsync(string[] args)
    {
        if (!TryParse(args, out var provider, out var connection, out var migrationsPath, out var showHelp, out var error))
            return showHelp ? WriteHelp() : Fail(error);

        if (!MigrateCli.TryResolveProvider(dialects, connections, provider, connection, out var dialect, out connection, out error))
            return Fail(error);

        var resolvedMigrationsPath = migrationsPath ?? $"./sql/{dialect.Name}/migrations";

        try
        {
            await using var conn = dialect.CreateConnection(dialect.NormalizeConnectionString(connection));
            await conn.OpenAsync().ConfigureAwait(false);
            await DbExecute.NonQueryAsync(conn, null, dialect.MigratesDdl).ConfigureAwait(false);
            await DbExecute.NonQueryAsync(conn, null, dialect.MigrateLogsDdl).ConfigureAwait(false);
            Directory.CreateDirectory(resolvedMigrationsPath);
        }
        catch (Exception ex)
        {
            return Fail($"setup failed: {ex.Message}", 1);
        }

        Console.Write(
            HelpTemplates.Message(
                "setup-complete",
                new Dictionary<string, string>
                {
                    ["dialect"] = dialect.Name,
                    ["migrationsPath"] = resolvedMigrationsPath,
                    ["createExample"] =
                        $"  migrate-create --provider {dialect.Name} --name add_users",
                    ["upExample"] =
                        $"  migrate-up --provider {dialect.Name} --connection {connection}",
                }
            )
        );
        return 0;
    }

    private bool TryParse(
        string[] args,
        out string provider,
        out string? connection,
        out string? migrationsPath,
        out bool showHelp,
        out string error
    )
    {
        provider = string.Empty;
        connection = null;
        migrationsPath = null;
        showHelp = false;
        error = string.Empty;

        for (var i = 0; i < args.Length; i++)
        {
            var a = args[i];
            if (a is "-h" or "--help")
            {
                showHelp = true;
                return false;
            }
            if (MigrateCli.TryTakeFlag(args, ref i, a, "--provider", out provider))
                continue;
            if (MigrateCli.TryTakeFlag(args, ref i, a, "--connection", out connection))
                continue;
            if (MigrateCli.TryTakeFlag(args, ref i, a, "--migrations-path", out migrationsPath)
                || MigrateCli.TryTakeFlag(args, ref i, a, "--migrate-path", out migrationsPath))
                continue;

            error = MigrateCli.ParseArgError(
                a,
                a.StartsWith("--", StringComparison.Ordinal) && i + 1 >= args.Length
            );
            return false;
        }

        return !string.IsNullOrEmpty(provider) || MigrateCli.MissingValue("--provider", out error);
    }

    private static int WriteHelp()
    {
        Console.Error.Write(HelpText);
        return 0;
    }

    private static int Fail(string error, int code = 2)
    {
        Console.Error.WriteLine(error);
        if (code == 2)
            Console.Error.Write(HelpText);
        return code;
    }
}

File: rust/src/bin/migrate_setup.rs
// Idempotent CREATE TABLE IF NOT EXISTS migrates+migrate_logs; mirrors scripts/lib/datasource-migrate.mjs:setupSql so rust + TS containers share one migrates contract.

use std::env;
use std::fs;
use std::path::PathBuf;
use std::process;

use deterministic_migraters::{
    connection_from_env, fill_help_template, message, try_get, MigrationPool, SqlDialect,
    SUPPORTED_PROVIDERS,
};

const HELP_TEMPLATE: &str = include_str!("../../../shared/templates/help/setup.txt");

fn help_text() -> String {
    fill_help_template(HELP_TEMPLATE, "migrate-setup")
}

struct Args {
    dialect: &'static dyn SqlDialect,
    connection: String,
    migrations_path: String,
    and_up: bool,
}

fn parse_args() -> Result<Args, String> {
    let mut provider: Option<String> = None;
    let mut connection: Option<String> = None;
    let mut migrations_path: Option<String> = None;
    let mut and_up = false;
    let mut iter = env::args().skip(1);
    while let Some(a) = iter.next() {
        match a.as_str() {
            "--provider" => provider = iter.next(),
            "--connection" => connection = iter.next(),
            "--migrations-path" => migrations_path = iter.next(),
            "--migrate-path" => migrations_path = iter.next(),
            "--and-up" => and_up = true,
            "-h" | "--help" => {
                eprint!("{}", help_text());
                process::exit(0);
            }
            other => return Err(message("errors/unknown-arg", &[("arg", other)])),
        }
    }
    let Some(provider) = provider else {
        return Err(message(
            "errors/missing-provider-hint",
            &[("providers", SUPPORTED_PROVIDERS)],
        ));
    };
    let Some(dialect) = try_get(&provider) else {
        return Err(message(
            "errors/unsupported-provider",
            &[("provider", &provider)],
        ));
    };
    let connection = match connection.or_else(|| connection_from_env(dialect)) {
        Some(c) => c,
        None => {
            return Err(message(
                "errors/missing-connection",
                &[("envVars", &dialect.connection_env_vars().join(", "))],
            ))
        }
    };
    let migrations_path =
        migrations_path.unwrap_or_else(|| format!("./sql/{}/migrations", dialect.name()));
    Ok(Args {
        dialect,
        connection,
        migrations_path,
        and_up,
    })
}

fn sibling_migrate_up_binary() -> Result<PathBuf, Box<dyn std::error::Error>> {
    let me = env::current_exe()?;
    let dir = me.parent().ok_or("current_exe has no parent")?;
    let mut name = String::from("migrate-up");
    if let Some(ext) = me.extension().and_then(|e| e.to_str()) {
        name.push('.');
        name.push_str(ext);
    }
    Ok(dir.join(name))
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    match dotenvy::from_path(".env") {
        Ok(()) => {}
        Err(dotenvy::Error::Io(err)) if err.kind() == std::io::ErrorKind::NotFound => {}
        Err(err) => return Err(format!("failed to load .env: {err}").into()),
    }
    let args = match parse_args() {
        Ok(a) => a,
        Err(msg) => {
            eprintln!("{}", msg);
            process::exit(2);
        }
    };
    args.dialect.prepare_setup(&args.connection)?;
    let pool = MigrationPool::connect(args.dialect, &args.connection).await?;
    pool.execute(args.dialect.migrates_ddl()).await?;
    pool.execute(args.dialect.migrate_logs_ddl()).await?;
    fs::create_dir_all(PathBuf::from(&args.migrations_path))?;
    print!(
        "{}",
        message(
            "setup-complete",
            &[
                ("dialect", args.dialect.name()),
                ("migrationsPath", &args.migrations_path),
                (
                    "createExample",
                    &format!(
                        "  cargo run --release --bin migrate-create -- --provider {} --name add_users",
                        args.dialect.name()
                    ),
                ),
                (
                    "upExample",
                    &format!(
                        "  cargo run --release --bin migrate-up -- --provider {} --connection {}",
                        args.dialect.name(),
                        args.connection
                    ),
                ),
            ],
        )
    );
    if args.and_up {
        let up = sibling_migrate_up_binary()?;
        let status = std::process::Command::new(&up)
            .arg("--provider")
            .arg(args.dialect.name())
            .arg("--connection")
            .arg(&args.connection)
            .arg("--migrations-path")
            .arg(&args.migrations_path)
            .status()?;
        if !status.success() {
            process::exit(status.code().unwrap_or(1));
        }
    }
    Ok(())
}
